File: js/src/handles.ts
/**
 * Trace and Step Handles
 * User-facing objects for managing trace/step lifecycle
 */

import { v4 as uuidv4 } from 'uuid';
import {
  TraceHandle,
  StepHandle,
  TraceEvent,
  TraceEventType,
  FinishTraceOptions,
  FailTraceOptions,
  FinishStepOptions,
  FailStepOptions,
  StartStepOptions,
  LogOptions,
  LogLevel,
} from './types';
import { ContextManager } from './context-manager';
import { LoggerLike } from './logger';
import { createTraceEvent } from './event-factory';
import { ensureValidUuid } from './validate-uuid';

/**
 * Internal trace handle implementation
 */
export class TraceHandleImpl implements TraceHandle {
  public readonly trace_id: string;
  private source: string;
  private sendEvent: (event: TraceEvent) => Promise<void>;
  private contextManager: ContextManager;
  private logger: LoggerLike;
  private closed: boolean = false;
  private steps: StepHandleImpl[] = [];
  private onClose?: () => void;

  constructor(
    trace_id: string,
    source: string,
    sendEvent: (event: TraceEvent) => Promise<void>,
    contextManager: ContextManager,
    logger?: LoggerLike,
    onClose?: () => void,
  ) {
    this.trace_id = trace_id;
    this.source = source;
    this.sendEvent = sendEvent;
    this.contextManager = contextManager;
    const noop = () => {};
    this.logger = logger || { debug: noop, info: noop, warn: noop, error: noop };
    this.onClose = onClose;
  }

  isClosed(): boolean {
    return this.closed;
  }

  /**
   * Finish trace successfully
   */
  async finish(options?: FinishTraceOptions): Promise<void> {
    if (this.closed) {
      this.logger.warn(`Trace ${this.trace_id} already closed`);
      return;
    }

    await this.closeOrphanedSteps('Parent trace finished');
    this.closed = true;
    this.onClose?.();

    await this.sendEvent(createTraceEvent(
      TraceEventType.TRACE_FINISHED,
      this.trace_id,
      this.source,
      { result: options?.result, metadata: options?.metadata },
    ));
  }

  /**
   * Fail trace with error
   */
  async fail(error: string | Error, options?: FailTraceOptions): Promise<void> {
    if (this.closed) {
      this.logger.warn(`Trace ${this.trace_id} already closed`);
      return;
    }

    const errorMessage = error instanceof Error ? error.message : error;
    const errorStack = error instanceof Error ? error.stack : undefined;

    await this.closeOrphanedSteps(errorMessage);
    this.closed = true;
    this.onClose?.();

    await this.sendEvent(createTraceEvent(
      TraceEventType.TRACE_FAILED,
      this.trace_id,
      this.source,
      { error: errorMessage, stack: errorStack, result: options?.result, metadata: options?.metadata },
    ));
  }

  /**
   * Cancel trace
   */
  async cancel(): Promise<void> {
    if (this.closed) {
      this.logger.warn(`Trace ${this.trace_id} already closed`);
      return;
    }

    await this.closeOrphanedSteps('Parent trace cancelled');
    this.closed = true;
    this.onClose?.();

    await this.sendEvent(createTraceEvent(
      TraceEventType.TRACE_CANCELLED,
      this.trace_id,
      this.source,
      {},
    ));
  }

  /**
   * Start a new step within this trace
   */
  async startStep(options?: StartStepOptions): Promise<StepHandle> {
    const step_id = ensureValidUuid(options?.step_id, this.logger, 'step_id');

    await this.sendEvent(createTraceEvent(
      TraceEventType.STEP_STARTED,
      this.trace_id,
      this.source,
      {
        name: options?.name,
        step_type: options?.step_type,
        input: options?.input,
        metadata: options?.metadata,
      },
      step_id,
    ));

    // Update context with step_id
    this.contextManager.updateContext({ step_id });

    const step = new StepHandleImpl(
      step_id,
      this.trace_id,
      this.source,
      this.sendEvent,
      this.contextManager,
      this.logger,
    );

    this.steps.push(step);
    return step;
  }

  /**
   * Execute a callback within a step, guaranteeing the step is closed.
   */
  async withStep<T>(fn: (step: StepHandle) => Promise<T>, options?: StartStepOptions): Promise<T> {
    const step = await this.startStep(options);
    try {
      const result = await fn(step);
      await step.finish({ output: result });
      return result;
    } catch (e) {
      await step.fail(e as Error);
      throw e;
    }
  }

  /**
   * Log message for this trace
   */
  async log(message: string, options?: LogOptions): Promise<void> {
    await this.sendEvent(createTraceEvent(
      TraceEventType.LOG_EMITTED,
      this.trace_id,
      this.source,
      {
        message,
        level: options?.level || LogLevel.INFO,
        event_type: options?.event_type,
        details: options?.details,
      },
      options?.step_id,
    ));
  }

  private async closeOrphanedSteps(reason: string): Promise<void> {
    for (const step of this.steps) {
      if (!step.isClosed()) {
        try {
          await step.fail(reason);
        } catch {
          // ignore
        }
      }
    }
    this.steps = [];
  }
}

/**
 * Internal step handle implementation
 */
export class StepHandleImpl implements StepHandle {
  public readonly step_id: string;
  public readonly trace_id: string;
  private source: string;
  private sendEvent: (event: TraceEvent) => Promise<void>;
  private contextManager: ContextManager;
  private logger: LoggerLike;
  private closed: boolean = false;
  private onClose?: () => void;

  constructor(
    step_id: string,
    trace_id: string,
    source: string,
    sendEvent: (event: TraceEvent) => Promise<void>,
    contextManager: ContextManager,
    logger?: LoggerLike,
    onClose?: () => void,
  ) {
    this.step_id = step_id;
    this.trace_id = trace_id;
    this.source = source;
    this.sendEvent = sendEvent;
    this.contextManager = contextManager;
    const noop = () => {};
    this.logger = logger || { debug: noop, info: noop, warn: noop, error: noop };
    this.onClose = onClose;
  }

  isClosed(): boolean {
    return this.closed;
  }

  /**
   * Finish step successfully
   */
  async finish(options?: FinishStepOptions): Promise<void> {
    if (this.closed) {
      this.logger.warn(`Step ${this.step_id} already closed`);
      return;
    }

    this.closed = true;
    this.onClose?.();

    await this.sendEvent(createTraceEvent(
      TraceEventType.STEP_FINISHED,
      this.trace_id,
      this.source,
      { output: options?.output, metadata: options?.metadata },
      this.step_id,
    ));

    // Clear step from context
    this.contextManager.updateContext({ step_id: undefined });
  }

  /**
   * Fail step with error
   */
  async fail(error: string | Error, options?: FailStepOptions): Promise<void> {
    if (this.closed) {
      this.logger.warn(`Step ${this.step_id} already closed`);
      return;
    }

    this.closed = true;
    this.onClose?.();

    const errorMessage = error instanceof Error ? error.message : error;
    const errorStack = error instanceof Error ? error.stack : undefined;

    await this.sendEvent(createTraceEvent(
      TraceEventType.STEP_FAILED,
      this.trace_id,
      this.source,
      { error: errorMessage, stack: errorStack, output: options?.output, metadata: options?.metadata },
      this.step_id,
    ));

    // Clear step from context
    this.contextManager.updateContext({ step_id: undefined });
  }

  /**
   * Log message for this step
   */
  async log(message: string, options?: LogOptions): Promise<void> {
    await this.sendEvent(createTraceEvent(
      TraceEventType.LOG_EMITTED,
      this.trace_id,
      this.source,
      {
        message,
        level: options?.level || LogLevel.INFO,
        event_type: options?.event_type,
        details: options?.details,
      },
      this.step_id,
    ));
  }
}

File: go/handles.go
package traceflow

import (
	"context"
	"sync"
)

// TraceHandle manages the lifecycle of a single trace. It is safe for
// concurrent use. Finish, Fail, and Cancel are idempotent: the first call wins
// and subsequent calls are no-ops.
type TraceHandle struct {
	traceID string
	source  string
	emit    func(context.Context, TraceEvent) error
	logger  Logger
	onClose func()

	mu     sync.Mutex
	closed bool
	steps  []*StepHandle
}

// TraceID returns the trace's unique identifier.
func (h *TraceHandle) TraceID() string { return h.traceID }

// IsClosed reports whether the trace has been finished, failed, or cancelled.
func (h *TraceHandle) IsClosed() bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.closed
}

// markClosed transitions the handle to closed exactly once. It returns false if
// the handle was already closed.
func (h *TraceHandle) markClosed() bool {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.closed {
		h.logger.Warn("Trace %s already closed", h.traceID)
		return false
	}
	h.closed = true
	return true
}

// Finish completes the trace successfully, cascading closure to any open steps.
func (h *TraceHandle) Finish(ctx context.Context, opts FinishTraceOptions) error {
	if !h.markClosed() {
		return nil
	}
	h.closeOrphanedSteps(ctx, "Parent trace finished")
	if h.onClose != nil {
		h.onClose()
	}
	return h.emit(ctx, newTraceEvent(EventTraceFinished, h.traceID, h.source, map[string]any{
		"result":   opts.Result,
		"metadata": opts.Metadata,
	}, ""))
}

// Fail marks the trace as failed, cascading closure to any open steps.
func (h *TraceHandle) Fail(ctx context.Context, err error, opts FailTraceOptions) error {
	if !h.markClosed() {
		return nil
	}
	msg := ""
	if err != nil {
		msg = err.Error()
	}
	h.closeOrphanedSteps(ctx, msg)
	if h.onClose != nil {
		h.onClose()
	}
	return h.emit(ctx, newTraceEvent(EventTraceFailed, h.traceID, h.source, map[string]any{
		"error":    msg,
		"result":   opts.Result,
		"metadata": opts.Metadata,
	}, ""))
}

// Cancel marks the trace as cancelled, cascading closure to any open steps.
func (h *TraceHandle) Cancel(ctx context.Context) error {
	if !h.markClosed() {
		return nil
	}
	h.closeOrphanedSteps(ctx, "Parent trace cancelled")
	if h.onClose != nil {
		h.onClose()
	}
	return h.emit(ctx, newTraceEvent(EventTraceCancelled, h.traceID, h.source, map[string]any{}, ""))
}

// StartStep begins a new step within this trace. The step is tracked so it can
// be auto-closed if the trace ends while the step is still open.
func (h *TraceHandle) StartStep(ctx context.Context, opts StartStepOptions) (*StepHandle, error) {
	stepID := ensureValidUUID(opts.StepID, h.logger, "step_id")

	if err := h.emit(ctx, newTraceEvent(EventStepStarted, h.traceID, h.source, map[string]any{
		"name":      opts.Name,
		"step_type": opts.StepType,
		"input":     opts.Input,
		"metadata":  opts.Metadata,
	}, stepID)); err != nil {
		return nil, err
	}

	step := &StepHandle{
		stepID:  stepID,
		traceID: h.traceID,
		source:  h.source,
		emit:    h.emit,
		logger:  h.logger,
	}
	h.mu.Lock()
	h.steps = append(h.steps, step)
	h.mu.Unlock()
	return step, nil
}

// WithStep starts a step, runs fn, then finishes the step with fn's result on
// success or fails it on error. The error from fn is always propagated.
func (h *TraceHandle) WithStep(ctx context.Context, opts StartStepOptions, fn func(ctx context.Context, step *StepHandle) error) error {
	step, err := h.StartStep(ctx, opts)
	if err != nil {
		return err
	}
	stepCtx := ContextWithStep(ContextWithTrace(ctx, h.traceID), step.stepID)
	if runErr := fn(stepCtx, step); runErr != nil {
		_ = step.Fail(ctx, runErr, FailStepOptions{})
		return runErr
	}
	return step.Finish(ctx, FinishStepOptions{})
}

// Log emits a log line associated with this trace.
func (h *TraceHandle) Log(ctx context.Context, message string, opts LogOptions) error {
	level := opts.Level
	if level == "" {
		level = LogInfo
	}
	return h.emit(ctx, newTraceEvent(EventLogEmitted, h.traceID, h.source, map[string]any{
		"message":    message,
		"level":      level,
		"event_type": opts.EventType,
		"details":    opts.Details,
	}, opts.StepID))
}

func (h *TraceHandle) closeOrphanedSteps(ctx context.Context, reason string) {
	h.mu.Lock()
	steps := h.steps
	h.steps = nil
	h.mu.Unlock()

	reasonErr := stringError(reason)
	for _, step := range steps {
		if !step.IsClosed() {
			_ = step.Fail(ctx, reasonErr, FailStepOptions{})
		}
	}
}

// StepHandle manages the lifecycle of a single step. It is safe for concurrent
// use. Finish and Fail are idempotent.
type StepHandle struct {
	stepID  string
	traceID string
	source  string
	emit    func(context.Context, TraceEvent) error
	logger  Logger

	mu     sync.Mutex
	closed bool
}

// StepID returns the step's unique identifier.
func (s *StepHandle) StepID() string { return s.stepID }

// TraceID returns the parent trace's identifier.
func (s *StepHandle) TraceID() string { return s.traceID }

// IsClosed reports whether the step has been finished or failed.
func (s *StepHandle) IsClosed() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.closed
}

func (s *StepHandle) markClosed() bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		s.logger.Warn("Step %s already closed", s.stepID)
		return false
	}
	s.closed = true
	return true
}

// Finish completes the step successfully.
func (s *StepHandle) Finish(ctx context.Context, opts FinishStepOptions) error {
	if !s.markClosed() {
		return nil
	}
	return s.emit(ctx, newTraceEvent(EventStepFinished, s.traceID, s.source, map[string]any{
		"output":   opts.Output,
		"metadata": opts.Metadata,
	}, s.stepID))
}

// Fail marks the step as failed.
func (s *StepHandle) Fail(ctx context.Context, err error, opts FailStepOptions) error {
	if !s.markClosed() {
		return nil
	}
	msg := ""
	if err != nil {
		msg = err.Error()
	}
	return s.emit(ctx, newTraceEvent(EventStepFailed, s.traceID, s.source, map[string]any{
		"error":    msg,
		"output":   opts.Output,
		"metadata": opts.Metadata,
	}, s.stepID))
}

// Log emits a log line associated with this step.
func (s *StepHandle) Log(ctx context.Context, message string, opts LogOptions) error {
	level := opts.Level
	if level == "" {
		level = LogInfo
	}
	return s.emit(ctx, newTraceEvent(EventLogEmitted, s.traceID, s.source, map[string]any{
		"message":    message,
		"level":      level,
		"event_type": opts.EventType,
		"details":    opts.Details,
	}, s.stepID))
}

// stringError adapts a reason string to the error interface for step failures.
type stringError string

func (e stringError) Error() string { return string(e) }
